README
= FINANCE

a library for financial modelling in Ruby.

== INSTALL

  $ sudo gem install finance

== OVERVIEW

=== GETTING STARTED

  >> require 'finance'

*Note:* As of version 1.0.0, the entire library is contained under the
Finance namespace.  Existing code will not work unless you add:

  >> include Finance

for all of the examples below, we'll assume that you have done this.

=== AMORTIZATION

You are interested in borrowing $250,000 under a 30 year, fixed-rate
loan with a 4.25% APR.

  >> rate = Rate.new(0.0425, :apr, :duration => 30.years)
  >> amortization = 250000.amortize(rate)

Find the standard monthly payment:

  >> amortization.payment
  => DecNum('-1229.91')

Find the total cost of the loan:

  >> amortization.payments.sum
  => DecNum('-442766.55')

How much will you pay in interest?

  >> amortization.interest.sum
  => DecNum('192766.55')

How much interest in the first six months?

  >> amortization.interest[0,6].sum
  => DecNum('5294.62')

If your loan has an adjustable rate, no problem.  You can pass an
arbitrary number of rates, and they will be used in the amortization.
For example, we can look at an amortization of $250000, where the APR
starts at 4.25%, and increases by 1% every five years.

  >> values = %w{ 0.0425 0.0525 0.0625 0.0725 0.0825 0.0925 }
  >> rates = values.collect { |value| Rate.new( value, :apr, :duration = 5.years ) }
  >> arm = Amortization.new(250000, *rates)

Since we are looking at an ARM, there is no longer a single "payment" value.

  >> arm.payment
  => nil

But we can look at the different payments over time.

  >> arm.payments.uniq
  => [DecNum('-1229.85'), DecNum('-1360.41'), DecNum('-1475.65'), DecNum('-1571.07'), ... snipped ... ]

The other methods previously discussed can be accessed in the same way:

  >> arm.interest.sum
  => DecNum('287515.45')
  >> arm.payments.sum
  => DecNum('-537515.45')

Last, but not least, you may pass a block when creating an Amortization
which returns a modified monthly payment.  For example, to increase your
payment by $150, do:

  >> rate = Rate.new(0.0425, :apr, :duration => 30.years)
  >> extra_payments = 250000.amortize(rate){ |period| period.payment - 150 }

Disregarding the block, we have used the same parameters as the first
example.  Notice the difference in the results:

  >> amortization.payments.sum
  => DecNum('-442745.98')
  >> extra_payments.payments.sum
  => DecNum('-400566.24')
  >> amortization.interest.sum
  => DecNum('192745.98')
  >> extra_payments.interest.sum
  => DecNum('150566.24')

You can also increase your payment to a specific amount:

  >> extra_payments_2 = 250000.amortize(rate){ -1500 }

== Effective Interest Rate

Implementation of Execl's RATE function

  RATE  RATE(NPER, PMT, PV, FV, type, guess)  
  Returns the constant interest rate per period of an annuity. 
  => NPER is the total number of periods, during which payments are made (payment period). 
  => PMT is the constant payment (annuity) paid during each period. 
  => PV is the cash value in the sequence of payments. 
  => FV (optional) is the future value, which is reached at the end of the periodic payments. 
  => Type (optional) defines whether the payment is due at the beginning (1) or the end (0) of a period. 
  => Guess (optional) determines the estimated value of the interest with iterative calculation.
  see test/test_effective_interest_rate.rb for examples
  adapted from Java code found here: http://www.pcpros.com/software/programming/java/java.shtml
  Example:
  Finance::EffectiveInterestRate.calc_effective_interest_rate(360, -1433.39, 200000)

== Financial Payments with annual fees

This will return identical results that Finance::Amortization will given annual_fee = 0
Simply adds ability to use annual_fee

    Finance::EffectiveInterestRate.calc_payment(principal, rate, periods, annual_fee = 0)

== ABOUT

I started developing _finance_ while analyzing mortgages as a personal
project.  Spreadsheets have convenient formulas for doing this type of
work, until you want to do something semi-complex (like ARMs or extra
payments), at which point you need to create your own amortization
table.  I thought I could create a better interface for this type of
work in Ruby, and since I couldn't find an existing resource for these
tools, I am hoping to save other folks some time by releasing what I
have as a gem.

More broadly, I believe there are many calculations that are necessary
for the effective management of personal finances, but are difficult
(or impossible) to do with spreadsheets or other existing open source
tools.  My hope is that the _finance_ library will grow to provide a set
of open, tested tools to fill this gap.

If you have used _finance_ and find it useful, I would enjoy hearing
about it!

== FEATURES

Currently implemented features include:

* Uses the {http://flt.rubyforge.org/ flt} library to ensure precision decimal arithmetic in all calculations.
* Fixed-rate mortgage amortization (30/360).
* Interest rates
* Various cash flow computations, such as NPV and IRR.
* Adjustable rate mortgage amortization.
* Payment modifications (i.e., how does paying an additional $75 per month affect the amortization?)

== RESOURCES

[RubyGems Page] {https://rubygems.org/gems/finance}
[Source Code] {http://github.com/wkranec/finance}
[Bug Tracker] {https://github.com/wkranec/finance/issues}
[Google Group] {http://groups.google.com/group/finance-gem/topics?pli=1}

== COPYRIGHT

This library is released under the terms of the LGPL license.

Copyright (c) 2011, William Kranec.
All rights reserved.

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program.  If not, see <http://www.gnu.org/licenses/>.


HISTORY
= Version 2.0.1
9 May 2014

* Added Support for configuration file to set up default eps and guess for IRR & XIRR
* Added guess rate for IRR & XIRR
* NVP now does not change the given cashflow array

= Version 2.0.0
23 Jul 2013

* Removed Integer#months, Integer#years, and replaced Numeric#to_d by Numeric#to_s in the interest of Rails compatibility.
* Converted unit tests from the shoulda framework to minitest.
* Removed octal numbers in test_cashflow.rb
* Thanks to @thadd, @bramswenson, and @xpe for their contributions to this release!

= Version 1.1.2
16 Jun 2012

* Bugfix: Array#irr and Array#xirr check for a valid sequence of cash flows.
* Bugfix: Integer#months and Integer#years no longer collide with Rails methods.

= Version 1.1.0
11 Sep 2011

* Added XNPV and XIRR functions, with basic testing.
* Bugfix: Array#sum no longer collides with the Array#sum defined in Rails.
* Bugfix: Numeric#amortize now correctly calls Finance::Amortization#new.

= Version 1.0.0
20 Jul 2011

* Moved to Ruby 1.9.
* All classes are now contained within the +Finance+ namespace.
* LOTS of additional documentation and examples.
* Introduced _shoulda_ for unit tests, to make things a little more readable.
* Bugfix: The +amortize+ Numeric method now accepts a variable number of rates.
* Some code refactoring and clean-up for a small performance increase.

= Version 0.2.0
28 Jun 2011

* Added support for adjustable rate mortgages.
* Added support for additional payments.

= Version 0.1.1
21 Jun 2011

* Code examples in README now display correctly in the online documentation.

= Version 0.1.0
21 Jun 2011

* Support for fixed-rate mortgage amortization.
* NPV, IRR array methods for cash flow analysis.


lib/amortization.rb
require 'cashflows'
require 'transaction'

require 'rubygems'
require 'flt'

# the Amortization class provides an interface for working with loan amortizations.
# @example Borrow $250,000 under a 30 year, fixed-rate loan with a 4.25% APR
#   rate = Rate.new(0.0425, :apr, :duration => 30.years)
#   amortization = 250000.amortize(rate)
# @example Borrow $250,000 under a 30 year, adjustable rate loan, with an APR starting at 4.25%, and increasing by 1% every five years
#   values = %w{ 0.0425 0.0525 0.0625 0.0725 0.0825 0.0925 }
#   rates = values.collect { |value| Rate.new( value, :apr, :duration = 5.years ) }
#   arm = Amortization.new(250000, *rates)
# @example Borrow $250,000 under a 30 year, fixed-rate loan with a 4.25% APR, but pay $150 extra each month
#   rate = Rate.new(0.0425, :apr, :duration => 30.years)
#   extra_payments = 250000.amortize(rate){ |period| period.payment - 150 }
# @api public
class Amortization
  # @return [Numeric] the balance of the loan at the end of the amortization period (usually zero)
  # @api public
  attr_reader :balance
  # @return [Array] the interest charges on the loan
  # @api public
  attr_reader :interest
  # @return [Numeric] the required monthly payment.  For loans with more than one rate, returns nil
  # @api public
  attr_reader :payment
  # @return [Numeric] the principal amount of the loan
  # @api public
  attr_reader :principal
  # @return [Array] the interest rates used for calculating the amortization
  # @api public
  attr_reader :rates

  # @return [Numeric] -1, 0, or +1
  # @param [Amortization]
  # @api public
  def ==(amortization)
    self.principal == amortization.principal and self.rates == amortization.rates and self.payments == amortization.payments
  end

  # @return [Array] the amount of any additional payments in each period
  # @api public
  def additional_payments
    @pmts.collect{ |p| p.additional_amount }
  end

  # amortize the balance of loan with the given interest rate
  # @return none
  # @param [Rate] rate the interest rate to use in the amortization
  # @api private
  def amortize(rate)
    # For the purposes of calculating a payment, the relevant time
    # period is the remaining number of periods in the loan, not
    # necessarily the duration of the rate itself.
    periods = @rate_duration - @pmts.length
    amount = Amortization.payment @balance, rate.monthly, periods

    pmt = Transaction.new(amount)
    if @block then pmt.modify(&@block) end
      
    rate.duration.times do
      # Do this first in case the balance is zero already.
      if @balance.zero? then break end

      # Compute and record interest on the outstanding balance.
      interest = (@balance * rate.monthly).round(2)
      @balance += interest
      @interest << interest

      # Don't pay more than the outstanding balance
      if pmt.amount.abs > @balance then pmt.amount = -@balance end

      @pmts << pmt.dup
      @balance += pmt.amount
    end
  end

  # compute the amortization of the principal
  # @return none
  # @api private
  def compute
    @balance = @principal
    @interest = []
    @pmts = []

    @rates.each do |rate|
      amortize(rate)
    end

    # Add any remaining balance due to rounding error to the last payment.
    unless @balance.zero?
      @pmts[-1].amount -= @balance
      @balance = 0
    end

    if @rates.length == 1
      @payment = self.payments[0]
    else
      @payment = nil
    end
  end

  # @return [Integer] the time required to pay off the loan, in months
  # @example
  #   rate = 
  # @api public
  def duration
    @pmts.length
  end

  # create a new Amortization instance
  # @return [Amortization]
  # @param [DecNum] principal the initial amount of the loan or investment
  # @param [Rate] rates the applicable interest rates
  # @param [Proc] block
  # @example there are two ways to create a new Amortization
  #   rate = 
  # @api public
  def initialize(principal, *rates, &block)
    @principal = principal
    @rates     = rates
    @rate_duration  = (rates.collect { |r| r.duration }).sum
    @block     = block

    compute
  end

  # @api public
  def inspect
    "Amortization.new(#{@principal})"
  end

  # @return [DecNum] the periodic payment due on a loan
  # @param [DecNum] principal the initial amount of the loan or investment
  # @param [Rate] rate the applicable interest rate (per period)
  # @param [Integer] periods the number of periods needed for repayment
  # @note in most cases, you will probably want to use rate.monthly when calling this function outside of an Amortization instance.
  # @example
  #   rate = Rate.new(0.0375, :apr, :duration => 30.years)
  #   rate.duration #=> 360
  #   Amortization.payment(200000, rate.monthly, rate.duration) #=> DecNum('-926.23')
  # @see http://en.wikipedia.org/wiki/Amortization_calculator
  # @api public
  def Amortization.payment(principal, rate, periods)
    -(principal * (rate + (rate / ((1 + rate) ** periods - 1)))).round(2)
  end

  # @return [Array] the amount of the payment in each period
  # @api public
  def payments
    @pmts.collect{ |p| p.amount }
  end
end

class Numeric
  # @see Amortization#new
  # @api public
  def amortize(*rates, &block)
    amortization = Amortization.new(self, *rates, &block)
  end
end
